src/app/modules/item/pages/item-display/item-display.component.ts

```diff
@@ -15,165 +22,160 @@ interface TaskTab {
-    this.task = task;
-    const taskParams = { minScore: -3, maxScore: 10, randomSeed: 0, noScore: 0, readOnly: false, options: {} };
-    this.platform = new ItemDisplayPlatform(task, taskParams);
-    this.task.bindPlatform(this.platform);
-
-    const initialViews = { task: true, solution: true, editor: true, hints: true, grader: true, metadata: true };
-    task.load(initialViews).subscribe(() => this.taskLoaded());
+  ngOnDestroy(): void {
+    this.subscriptions.forEach(s => s.unsubscribe());
+
+    // destroy the task
+    const state = this.state$.value;
+    if (state.isReady) state.data.destroy();
+  }
+
+  /** Initializes a task once the URL has been loaded in the iframe */
+  private loadTaskFromIframe(iframe: HTMLIFrameElement): void {
+    taskProxyFromIframe(iframe).pipe(
+      switchMap(task => {
+        // Got task proxy from the iframe, ask task to load
+        task.bindPlatform(this);
+        const initialViews = { task: true, solution: true, editor: true, hints: true, grader: true, metadata: true };
+        return task.load(initialViews).pipe(mapTo(task));
+      })
+    ).subscribe({
+      next: task => this.state$.next(readyState(task)),
+      error: err => this.state$.next(errorState(err))
+    });
   }
 
-  // task.load done
-  taskLoaded(): void {
-    this.state = 'loaded';
-
-    this.updateHeight();
-    this.heightInterval = interval(1000).subscribe(() => this.updateHeight());
-    this.saveInterval = interval(1000).subscribe(() => this.getAnswerState());
+  /* Communication with the task */
 
-    this.task?.showViewsInTask({ task: true }).subscribe(() => {});
-
-    this.task?.getViews().subscribe((views : any) => {
-      this.setViews(views);
+  private registerTaskViewLoading(): Subscription {
+    // Load views once the task has been loaded
+    return this.state$.pipe(
+      readyData(),
+      switchMap(task => task.getViews().pipe(map(views => ({ task, views })))),
+      tap(({ task: _, views }) => this.setTaskViews(views)),
+      switchMap(({ task, views: _ }) => task.showViewsInTask({ task: true })),
+    ).subscribe({
+      error: err => this.state$.next(errorState(err))
     });
   }
 
-  // Communication with the task
-  getAnswerState(): void {
-    if (!this.task) {
-      return;
-    }
-    forkJoin([
-      this.task.getAnswer(),
-      this.task.getState()
-    ]).subscribe(([ answer, state ]) => this.saveAnswerState(answer, state));
-  }
-
-  reloadAnswerState(answer : string, state : string): void {
-    if (!this.task) {
-      return;
-    }
-    const task = this.task;
-    task.reloadAnswer(answer)
-      .pipe(switchMap(() =>
-        task.reloadState(state)
-      ))
-      .subscribe();
+  private syncedHeight(): Observable<number> {
+    // Start updating the iframe height to match the task's height
+    return this.state$.pipe(
+      switchMap(s => interval(heightSyncInterval).pipe(mapTo(s))),
+      readyData(),
+      switchMap(task => task.getHeight()),
+    );
```

Comment:
private methods like this one are used directly as subscription, it would more "composable" to declare private observables directly and just subscribe to them

Reply:
Fixed